Refactor HSTSList to use HashMap

nmvk opened the pull request:
Refactored HSTSList to use HashMap, where the key of HashMap is the base domain. Every time when we check if a host is secure, we find the base domain of the host and get a vector of HSTS entries associated with the base domain. 

While this will not give O(1) look up time, we would have a smaller list to iterate for every lookup. I have added one unit test to validate `HashMap` changes.

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14756

<!-- Either: -->
- [X] There are tests for these changes OR

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14794)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/net/hsts.rs

emilio commented on components/net/hsts.rs:
```diff
@@ -78,40 +94,46 @@ impl HstsList {
-                e.matches_domain(host)
+        let base_domain = reg_host(host);
+
+        match self.entries_map.get(&base_domain) {
```
nit: You can use:

```rust
self.entries_map.get(&base_domain).map_or(false, |entries| {
   // ...
})
```

To save a bit of indentation. Also applies below.

mrnayak replied:
Thanks!

emilio commented on components/net/hsts.rs:
```diff
@@ -132,3 +148,7 @@ pub fn secure_url(url: &Url) -> Url {
     }
 }
+
+fn reg_host<'a>(url: &'a str) -> String {
```
I don't think this is needed? Just use `reg_suffix`, and when needed do `to_owned` or similar, but you don't need a `String` all the time.

mrnayak replied:
Thanks, I had to add `let host = entry.host.clone();` to solve borrowing issue is this fine?

emilio reviewed (approved):
Looks good to me once those comments are addressed, thanks for doing this! :)

nmvk committed:
Refactor HSTSList to use HashMap

Refactored HSTSList to use HashMap, where the key of HashMap is the base
domain. Every time when we check if a host is secure, we find the base
domain of the host and get a vector of HSTS entries associated with the
base domain.

While this will not give O(1) look up time, we would have a smaller list
to iterate for every lookup. I have added one unit test to validate
HashMap changes.

emilio commented:
@bors-servo r+

Yeah, it's ok.

Thanks for doing this! :)

bors-servo commented:
:pushpin: Commit 6a14349 has been approved by `emilio`

<!-- @bors-servo r=emilio 6a14349eb1611166220640c874eca389c93a5c40 -->

bors-servo commented:
:hourglass: Testing commit 6a14349 with merge b06f4ae...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4669), [arm32](http://build.servo.org/builders/arm32/builds/4840), [arm64](http://build.servo.org/builders/arm64/builds/4868), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4958), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1635), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1642), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4935), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4975), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1642), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1677), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/97), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/98)

mrnayak commented:
Thanks @emilio
